### PR TITLE
WIP - Tag remaining selectors as atomic

### PR DESCRIPTION
**Pull request by jpbetz**
#### What type of PR is this?

/kind cleanup

#### What this PR does / why we need it:

Without this change using server-side apply to change selectors can result in objects with unexpected selectors.

Based on a [review of all occurrences of the term "selector" in types.go files](https://docs.google.com/document/d/1jA39N3uvyDHheE-g8nLKbbnKkB3zhe4plAyl8qSdk_k/edit#).

TODO:
- [ ] Add tests

#### Which issue(s) this PR fixes:

Fixes #https://github.com/kubernetes/kubernetes/issues/97970

#### Special notes for your reviewer:

Related to https://github.com/kubernetes/kubernetes/pull/97989 but for 17 selectors found after reviewing all occurrences of the term "selector" in types.go files.

#### Does this PR introduce a user-facing change?

```release-note
Server Side Apply now treats all <*>Selector fields as atomic (meaning the entire selector is managed by a single writer and updated together), since they contain interrelated and inseparable fields that do not merge in intuitive ways.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

Please use the following format for linking documentation:
- KEP: https://github.com/kubernetes/enhancements/tree/master/keps/sig-api-machinery/555-server-side-apply
- Feature issue: https://github.com/kubernetes/enhancements/issues/555
- List of "selector" occurrences in types.go files: https://docs.google.com/document/d/1jA39N3uvyDHheE-g8nLKbbnKkB3zhe4plAyl8qSdk_k/edit

/sig api-machinery
/wg api-expression

### CODE REVIEW

File: staging/src/k8s.io/api/core/v1/types.go [hunk 1]
```diff
@@ -2657,6 +2662,7 @@ const (
 // The requirements of them are ANDed.
 // It provides a subset of functionality as NodeSelectorTerm.
 // This is an alpha feature and may change in the future.
+// +structType=atomic
 type TopologySelectorTerm struct {
```
Comment (liggitt) [inline staging/src/k8s.io/api/core/v1/types.go:2666]: most uses of this are in a list... should those lists be marked listType=atomic or is that the default?

Comment (jpbetz) [reply staging/src/k8s.io/api/core/v1/types.go:2666]: Good catch, those lists need to be marked as atomic. I'll fix.

Comment (jpbetz) [reply staging/src/k8s.io/api/core/v1/types.go:2666]: Actually, it should be the default. Verifying.

File: staging/src/k8s.io/api/core/v1/types.go [hunk 2]
```diff
@@ -2993,6 +3001,7 @@ type PodSpec struct {
 	// Selector which must match a node's labels for the pod to be scheduled on that node.
 	// More info: https://kubernetes.io/docs/concepts/configuration/assign-pod-node/
 	// +optional
+	// +mapType=atomic
```
Comment (Danil-Grigorev) [inline staging/src/k8s.io/api/core/v1/types.go:3004]: This is covered by https://github.com/kubernetes/kubernetes/pull/97989/files#diff-5ffbebb40a29f8e491f4ff1e07c618eb3edcc0d5f00a1a71445272d6767d7883R2985 and has included tests. No need to include this in that PR.

File: staging/src/k8s.io/api/core/v1/types.go [hunk 3]
```diff
@@ -3773,6 +3780,7 @@ type ReplicationControllerSpec struct {
 	// controller, if empty defaulted to labels on Pod template.
 	// More info: https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/#label-selectors
 	// +optional
+	// +mapType=atomic
```
Comment (Danil-Grigorev) [inline staging/src/k8s.io/api/core/v1/types.go:3785]: This is covered by https://github.com/kubernetes/kubernetes/pull/97989/

File: staging/src/k8s.io/api/core/v1/types.go [hunk 4]
```diff
@@ -4071,6 +4079,7 @@ type ServiceSpec struct {
 	// Ignored if type is ExternalName.
 	// More info: https://kubernetes.io/docs/concepts/services-networking/service/
 	// +optional
+	// +mapType=atomic
```
Comment (Danil-Grigorev) [inline staging/src/k8s.io/api/core/v1/types.go:4084]: This is covered by #97989

File: staging/src/k8s.io/api/core/v1/types.go (hunk 1 above)
Comment (jpbetz) [reply staging/src/k8s.io/api/core/v1/types.go:2666]: From the docs: "atomic and set apply to lists with scalar elements only" (https://kubernetes.io/docs/reference/using-api/server-side-apply/). 

@apelisse do you know if this is accurate?

File: staging/src/k8s.io/api/core/v1/types.go (hunk 3 above)
Comment (jpbetz) [reply staging/src/k8s.io/api/core/v1/types.go:3785]: Yes, I'll drop everything already in #97989 from this PR. Probably makes sense for #97989 to cover all the map typed fields.

File: staging/src/k8s.io/api/core/v1/types.go (hunk 1 above)
Comment (Jefftree) [reply staging/src/k8s.io/api/core/v1/types.go:2666]: @jpbetz: Discussed this with @apelisse today, lists are atomic by default and we shouldn't need to do anything extra for them. The documentation is around set lists which aren't applicable in our case (I'll make a todo to update our docs)

**Comment by apelisse** [comment]
AFAIU, we can close this one now!

**Comment by jpbetz** [comment]
Yep, this is superseded by the work @Jefftree is doing. Closed.

**Pull request closed without merging**